### PR TITLE
Effects Serialization

### PR DESCRIPTION
**Pull request by davepl**
Here’s how I’ve gone about things, will create a PR for it right away….

EffectsManager serializes like so:

inline void to_json(nlohmann::json& j, const IEffectsManager& manager) 
{
    // TODO - Serialize the effects, being sure to call the serializer that belongs
    // to the specific derviced effect type.

    j = {
        {"type", "EffectsManager"},
        {"fps", manager.GetFPS()},
        {"currentEffectIndex", manager.GetCurrentEffect()},
        {"effects", manager.Effects()}
    };
}

Of course to serialize it’s Effects(), it needs to be able to serialize an effect through its ILEDEffect interface, which is done thusly:

inline void to_json(nlohmann::json& j, const ILEDEffect& effect) 
{
    // Attempt to dynamically cast to each known concrete type and use its to_json handler
    
    if (const auto* colorWave = dynamic_cast<const ColorWaveEffect*>(&effect)) 
        to_json(j, *colorWave);
    else if (const auto* fireworks = dynamic_cast<const FireworksEffect*>(&effect))
        to_json(j, *fireworks);
    else if (const auto* solidColor = dynamic_cast<const SolidColorFill*>(&effect))
        to_json(j, *solidColor);
    else if (const auto* palette = dynamic_cast<const PaletteEffect*>(&effect))
        to_json(j, *palette);
    else if (const auto* starfield = dynamic_cast<const StarfieldEffect*>(&effect))
        to_json(j, *starfield);
    else if (const auto* video = dynamic_cast<const MP4PlaybackEffect*>(&effect))
        to_json(j, *video);
    else
        throw std::runtime_error("Unknown effect type for serialization");
}

And the other direction:

inline void from_json(const nlohmann::json& j, std::unique_ptr<ILEDEffect>& effect) 
{
    std::string type = j.at("type").get<std::string>();

    if (type == "ColorWave") {
        std::unique_ptr<ColorWaveEffect> temp;
        from_json(j, temp);
        effect = std::move(temp);
    } 
    else if (type == "Fireworks") {
        std::unique_ptr<FireworksEffect> temp;
        from_json(j, temp);
        effect = std::move(temp);
    } … etc …

And then of course I wrote the ones to go with each effect, which look like this:

inline void to_json(nlohmann::json& j, const ColorWaveEffect & effect) 
{
    j = {
        {"type", "ColorWave"},
        {"name", effect.Name()},
        {"speed", effect._speed},
        {"waveFrequency", effect._waveFrequency}
    };
}

inline void from_json(const nlohmann::json& j, unique_ptr<ColorWaveEffect>& effect) 
{
    effect =  make_unique<ColorWaveEffect>(
        j.at("name").get<string>(),
        j.value("speed", 0.5),
        j.value("waveFrequency", 10.0)
    );
}

I could static cast after losing at the “type” field, but kind of like this approach, since RTTI is on by default anyway!

### CODE REVIEW

File: effectsmanager.h [hunk 1]
```diff
@@ -190,5 +216,86 @@ class EffectsManager : public IEffectsManager
     {
         return _currentEffectIndex >= 0 && _currentEffectIndex < static_cast<int>(_effects.size());
     }
+
+    friend void to_json(nlohmann::json& j, const EffectsManager& manager);
+    friend void from_json(const nlohmann::json& j, EffectsManager& manager);
 };
 
+
+inline void to_json(nlohmann::json& j, const ILEDEffect& effect)
```
Comment (rbergen) [inline effectsmanager.h:225]: This type of repetitive code (and the corresponding code in the from_json() below it) make my fingers itch to "macro" it, but... I'll refrain. For now. :)

**Review by rbergen — APPROVED** [review]
Wow, I love what you did with the one comment I made about the challenge of polymorphic (de)serialization. Not many people can get a point like that that quickly, let alone immediately translate it to such a clean implementation in code!

Very nice, Dave!

File: effectsmanager.h (hunk 1 above)
Comment (davepl) [reply effectsmanager.h:235]: Consider snapping yourself with a rubber band each time you write a macro, it'll reduce the urge :-)

Comment (rbergen) [reply effectsmanager.h:235]: That's about as sensible as suggesting I punch myself in the face every time I empty the dishwasher. So no, I don't think I will. ;)

File: controller.h [hunk 2]
```diff
@@ -508,7 +508,11 @@ inline void from_json(const nlohmann::json &j, unique_ptr<Controller> & ptrContr
 
         // Extract canvases
         for (const auto &canvasJson : j.at("canvases"))
-            ptrController->AddCanvas(canvasJson.get<unique_ptr<ICanvas>>());
+        {
+            unique_ptr<ICanvas> ptrCanvas;
+            from_json(canvasJson, ptrCanvas);
+            ptrController->AddCanvas(std::move(ptrCanvas));
+        }
```
Comment (rbergen) [inline controller.h:515]: Is there a reason you prefer the more verbose way of doing this, and maybe even more so, specifically in this place?

**Comment by davepl** [comment]
Because macros are the devil’s playthings.  You can’t see through them, the IDE doesn’t expand them, you can’t read them easily, etc.  There are very few macros that can’t be inline functions.  

They’re one step below regular expressions and one step above AWK scripts in the “makes life harder” category.

I’ll see if I can craft some kind of clever inline function, and then you can still walk it in the debugger :-)

- Dave

**Comment by rbergen** [comment]
> Because macros are the devil’s playthings. [...]

Just for the record: you're responding to a comment about an entirely different thing. :) The comment I made about verbosity concerned you replacing this: 

```c++
        canvas->AddFeature(featureJson.get<unique_ptr<ILEDFeature>>());
```

with this:

```c++
        {
            unique_ptr<ILEDFeature> ptrFeature;
            from_json(featureJson, ptrFeature);
            canvas->AddFeature(std::move(ptrFeature));
        }
```

in Canvas's from_json(). They both do exactly the same thing.

Then back to your statement about macros: I get why they irk you, even though I have a different opinion - which is fine, we all have our preferences. But there is one thing I feel have to correct you on.

> You can’t see through them, the IDE doesn’t expand them, you can’t read them easily, etc.

Nowadays, the IDE (at least VSCode) does expand them when you hover your mouse over them. It even flags most compile-time errors in the expanded application of them. I admit getting to the root cause of the problem in the combination of the macro's definition and parameter expansion can be a puzzle, but still.

Anyway, I already said I've accepted your bad feelings towards macros in what is your project, so I won't introduce them.

**Comment by davepl** [comment]
Check out the current implementation, hopefully you like it better!  It’s checked into my branch…

- Dave

template<typename T>
unique_ptr<ILEDEffect> effectFactory(const nlohmann::json& j) {
    unique_ptr<T> effect;
    from_json(j, effect);
    return effect;
}

// Dynamically deserialize an effect from JSON based on its indicated type 
// and return it on the unique pointer out reference

inline void from_json(const nlohmann::json& j, unique_ptr<ILEDEffect>& effect) 
{
    static const std::unordered_map<std::string, std::unique_ptr<ILEDEffect>(*)(const nlohmann::json&)> effects_map = 
    {
        { ColorWaveEffect::EffectTypeName(),   effectFactory<ColorWaveEffect>   },
        { FireworksEffect::EffectTypeName(),   effectFactory<FireworksEffect>   },
        { SolidColorFill::EffectTypeName(),    effectFactory<SolidColorFill>    },
        { PaletteEffect::EffectTypeName(),     effectFactory<PaletteEffect>     },
        { StarfieldEffect::EffectTypeName(),   effectFactory<StarfieldEffect>   },
        { MP4PlaybackEffect::EffectTypeName(), effectFactory<MP4PlaybackEffect> }
    };
    auto it = effects_map.find(j["type"]);
    if (it == effects_map.end())
        throw runtime_error("Unknown effect type for deserialization: " + j["type"].get<string>());       
    effect = std::move(it->second(j));
}

**Comment by rbergen** [comment]
> Check out the current implementation, hopefully you like it better! It’s checked into my branch… 

Yeah, that's very clean. I definitely won't argue with that way of fixing it. 

By the way, as your branch is the one PR #17 is opened from anything you check into it ends up in there straight away; a PR always tracks the branch it's opened from, until it's merged.

File: controller.h (hunk 2 above)
Comment (rbergen) [reply controller.h:515]: This one too...

```suggestion
            ptrController->AddCanvas(canvasJson.get<unique_ptr<ICanvas>>());
```

**Review by rbergen — APPROVED** [review]
One more brevity suggestion. Based on you making a similar change earlier today I will apply it myself.

**Comment by rbergen** [comment]
CI is happy and so am I, so I'm merging this. Very nice job on the effects (de)serialization!

**Pull request merged**